### PR TITLE
[FEATURE] Backend work for usage tracking

### PR DESCRIPTION
**Pull request by sinkingpoint**
# Description

Linked with #1170

Here we add the backend side of usage tracking. We introduce a new view service that will be used to
track the usage of the API, bound to the /api/v1/view endpoint.

We then add a new implementation of that service
that tracks views in Prometheus metrics, with the
flexibility to swap out the implementation later.

# Screenshots

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

### CODE REVIEW

**Comment by sinkingpoint** [comment]
Opening this up. The only bit I'm slightly unsure on here is the `CheckPermissions` change in the toolbox. It works for my purposes, but there might be a better way

File: internal/api/impl/v1/view/endpoint.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	view := v1.View{}
+	if err := ctx.Bind(&view); err != nil {
+		return apiInterface.HandleBadRequestError(err.Error())
+	}
+
+	if err := view.Validate(); err != nil {
+		return apiInterface.HandleBadRequestError(err.Error())
+	}
+
+	if err := e.toolbox.CheckPermission(ctx, nil, apiInterface.Parameters{
```
Comment (celian-garcia) [inline internal/api/impl/v1/view/endpoint.go:56]: As @Gladorme  suggested me, it's simpler to use directly RBAC service actually 

Here is how the code would looks like 
```go
  claims := crypto.ExtractJWTClaims(ctx)
	if claims == nil {
		// Claims can be nil with anonymous endpoint and unauthenticated users, no need to continue
		return nil
	}
	if ok := e.rbac.HasPermission(claims.Subject, role.ReadAction, view.Project, role.DashboardScope); !ok {
		return apiInterface.HandleUnauthorizedError(fmt.Sprintf("missing '%s' permission in '%s' project for '%s' kind", role.ReadAction, view.Project, role.DashboardScope))
	}
```

Comment (celian-garcia) [reply internal/api/impl/v1/view/endpoint.go:56]: The CheckPermissions from toolbox has a lot of complicate conditions that doesn't apply with your need

**Review by celian-garcia — CHANGES_REQUESTED** [review]

File: internal/api/impl/v1/view/endpoint.go (hunk 1 above)
Comment (sinkingpoint) [reply internal/api/impl/v1/view/endpoint.go:56]: Nice. That's a lot cleaner - thanks!

File: internal/api/impl/v1/view/endpoint_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/perses/perses/internal/api/crypto"
+	apiInterface "github.com/perses/perses/internal/api/interface"
+	"github.com/perses/perses/internal/api/rbac"
+	"github.com/perses/perses/pkg/model/api/v1/role"
+	promclient "github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var _ = rbac.RBAC(&testRBAC{})
```
Comment (celian-garcia) [inline internal/api/impl/v1/view/endpoint_test.go:33]: Interesting, is this a way of testing that you copied from another place in Perses ? I would have used mockgen but I'm wondering what is the commonly used way since I didn't create some yet

File: pkg/model/api/v1/view.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+	if v.Dashboard == "" {
+		return fmt.Errorf("dashboard cannot be empty")
+	}
+
+	if v.RenderErrors < 0 {
+		return fmt.Errorf("render_errors cannot be negative")
+	}
+
+	if v.RenderTimeSecs < 0 {
```
Comment (celian-garcia) [inline pkg/model/api/v1/view.go:41]: Shouldn't we use a more precise unit ? Like ms ?

Comment (celian-garcia) [reply pkg/model/api/v1/view.go:41]: ah ok it's a float then why not

File: internal/api/impl/v1/view/service.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"github.com/perses/perses/internal/api/interface/v1/view"
+	v1 "github.com/perses/perses/pkg/model/api/v1"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// A counter for the total number of views of a dashboard.
+var dashboardViewCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+	Name: "perses_dashboard_view_total",
```
Comment (celian-garcia) [inline internal/api/impl/v1/view/service.go:24]:
```suggestion
	Name: "perses_dashboard_views_total",
```

File: internal/api/impl/v1/view/service.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	Help: "The total number of views of a dashboard",
+}, []string{"project", "dashboard"})
+
+// A counter for the total number of render errors of a dashboard.
+var dashboardRenderErrorCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+	Name: "perses_dashboard_render_error_total",
```
Comment (celian-garcia) [inline internal/api/impl/v1/view/service.go:30]:
```suggestion
	Name: "perses_dashboard_render_errors_total",
```

**Review by celian-garcia — APPROVED** [review]
Overall LGTM little question and suggestions. 

Maybe we could have a mechanism to enable it or not by config but it can be discussed apart.

File: pkg/model/api/v1/view.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	// The resource being viewed.
+	Project   string `json:"project" yaml:"project"`
+	Dashboard string `json:"dashboard" yaml:"dashboard"`
+
+	// Stats about the view.
+	RenderTimeSecs float64 `json:"render_time" yaml:"render_time"`
+	RenderErrors   int     `json:"render_errors" yaml:"render_errors"`
+}
+
+func (v *View) Validate() error {
```
Comment (Nexucis) [inline pkg/model/api/v1/view.go:28]: I would override the `Unmarshal` method for JSON and YAML to call this method. Like that when the struct is created from the body, we are sure the data are already good.

File: internal/api/impl/v1/view/service.go (hunk 4 above)
Comment (Nexucis) [inline internal/api/impl/v1/view/service.go:24]: Why not defining a constant `perses` for the metric namespace ?
We could also use it when defining the http server here: https://github.com/perses/perses/blob/main/internal/api/core/core.go#L50

File: internal/api/impl/v1/view/endpoint.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+		service: service,
+		rbac:    rbac,
+	}
+}
+
+func (e *endpoint) CollectRoutes(g *route.Group) {
+	g.POST(fmt.Sprintf("/%s", utils.PathView), e.View, false)
+}
+
+func (e *endpoint) View(ctx echo.Context) error {
```
Comment (Nexucis) [inline internal/api/impl/v1/view/endpoint.go:46]: I don't think you need to expose this method outside of this package

File: internal/api/dependency/service_manager.go [hunk 8]
```diff
@@ -102,6 +105,7 @@ type service struct {
 	secret             secret.Service
 	user               user.Service
 	variable           variable.Service
+	view               view.Service
```
Comment (Nexucis) [inline internal/api/dependency/service_manager.go:108]: from the issue, it mentioned an endpoint `visit` and here it's `view`. I don't really mind to change the name, just curious about the reason you changed it :).

File: internal/api/impl/v1/view/service.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+// scraped and stored by Prometheus.
+type metricsViewService struct {
+}
+
+func NewMetricsViewService() view.Service {
+	return &metricsViewService{}
+}
+
+func (m *metricsViewService) View(view *v1.View) error {
+	dashboardViewCounter.WithLabelValues(view.Project, view.Dashboard).Inc()
```
Comment (Nexucis) [inline internal/api/impl/v1/view/service.go:57]: Perhaps it's a good idea to check that the dashboard exists before changing the metrics. WDF ?

File: internal/api/impl/v1/view/service.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	Name: "perses_dashboard_render_error_total",
+	Help: "The total number of render errors of a dashboard",
+}, []string{"project", "dashboard"})
+
+// A histogram for the render time of a dashboard.
+var dashboardRenderTime = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name:    "perses_dashboard_render_time_seconds",
+	Help:    "The render time of a dashboard",
+	Buckets: prometheus.DefBuckets,
+}, []string{"project", "dashboard"})
```
Comment (Nexucis) [inline internal/api/impl/v1/view/service.go:39]: these labels are common for every metrics, it can be a good idea to put them as a constant

**Review by Nexucis — COMMENTED** [review]
nice, thanks for the PR. Overall like @celian-garcia it looks already good to me. I have added some proposition to improve it a bit.
Let me know wdyf.

File: internal/api/impl/v1/view/endpoint_test.go (hunk 2 above)
Comment (sinkingpoint) [reply internal/api/impl/v1/view/endpoint_test.go:33]: It's just a pattern to ensure that `testRBAC` implements the interface - https://go.dev/doc/faq#guarantee_satisfies_interface

File: internal/api/impl/v1/view/service.go (hunk 4 above)
Comment (sinkingpoint) [reply internal/api/impl/v1/view/service.go]: Done

File: pkg/model/api/v1/view.go (hunk 6 above)
Comment (sinkingpoint) [reply pkg/model/api/v1/view.go:66]: Done. Left one comment in the chat about this, but it seems to work

File: internal/api/impl/v1/view/service.go (hunk 4 above)
Comment (sinkingpoint) [reply internal/api/impl/v1/view/service.go]: Done. I've defined it as a namespace for these metrics - the actual flow of how we register metrics is a bit opaque, involving a magic import from perses/common, so we might want to rejig that flow in the future to make metrics registration a bit more explicit and consistent

File: internal/api/dependency/service_manager.go (hunk 8 above)
Comment (sinkingpoint) [reply internal/api/dependency/service_manager.go:108]: TBH, no particular reason. I completely forgot I named it visit in the issue :|

File: internal/api/impl/v1/view/service.go (hunk 9 above)
Comment (sinkingpoint) [reply internal/api/impl/v1/view/service.go:66]: Sounds reasonable. Done

File: internal/api/impl/v1/view/service.go (hunk 4 above)
Comment (Nexucis) [reply internal/api/impl/v1/view/service.go]: actually that shouldn't be that magic.

in `perses/common` we are providing default metrics (`_info` and metrics for the http server). Then it's simply possible to provide a prometheus metric registry to the `perses/common` app. If a metric registry is not provided, then the default one (provided by Prometheus) is used.

We are using a different metric registry in the e2e test: https://github.com/perses/perses/blob/main/internal/api/e2e/framework/server.go#L103

So today if we want to make metrics registration a bit more explicit and consistent we just need to define a new one and then register every metric in the `main` and then inject it in the `perses/common` app

File: pkg/model/api/v1/view.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+)
+
+var _ = json.Unmarshaler(&View{})
+var _ = yaml.Unmarshaler(&View{})
```
Comment (Nexucis) [inline pkg/model/api/v1/view.go:24]: we never needed to do that for other struct / resource definition. Why did you need it ?

Comment (sinkingpoint) [reply pkg/model/api/v1/view.go:24]: It's not _strictly_ necessary. Just a habit to show the intention that `View` implements those interfaces. I can remove it if you want

Comment (Nexucis) [reply pkg/model/api/v1/view.go:24]: oh I didn't know this trick. Interesting !
But yeah can you remove it as we don't do it across the whole code. (just to be consistent).

File: pkg/model/api/v1/view.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	if err := tmp.Validate(); err != nil {
+		return err
+	}
+	*v = tmp
+	return nil
+}
+
+func (v *View) Validate() error {
```
Comment (Nexucis) [inline pkg/model/api/v1/view.go:66]: and here looking at the rest of the code, you don't need to expose the method.

File: internal/api/impl/v1/view/service.go (hunk 4 above)
Comment (Nexucis) [reply internal/api/impl/v1/view/service.go]: I'm currently patching a bit `perses/common` that contains a bug about how we are using the prometheus registry.
And then I will integrate it in Perses. Let see if you find it more coherent !

**Review by Nexucis — APPROVED** [review]
once last comments are resolved, we should be good to merge !
Thank you for doing it @sinkingpoint !!
